src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -692,113 +663,69 @@ List<CompactionAggregate> getSelection(List<CompactionAggregate.UnifiedAggregate
                                            int runningAdaptiveCompactions,
                                            int maxAdaptiveCompactions)
     {
-        pending = controller.maybeSort(pending);
-
+        // Prepare parameters for the selection.
         int perLevelCount = totalCount / levelCount;   // each level has this number of tasks reserved for it
         int remainder = totalCount % levelCount;       // and the remainder is distributed randomly, up to 1 per level
 
-        // List the indexes of all compaction picks, adding several entries for compactions that span multiple shards.
-        IntArrayList list = new IntArrayList(pending.size(), -1);
-        IntArrayList expired = new IntArrayList(pending.size(), -1);
-        for (int aggregateIndex = 0; aggregateIndex < pending.size(); ++aggregateIndex)
+        // Calculate how many new ones we can add in each level, and how many we can assign randomly.
+        int remaining = totalCount;
+        for (int i = 0; i < levelCount; ++i)
+        {
+            remaining -= perLevel[i];
+            if (perLevel[i] > perLevelCount)
+                remainder -= perLevel[i] - perLevelCount;
+        }
+        // Note: if we are in the middle of changes in the parameters or level count, remainder might become negative.
```

Comment:
This also true when decreasing totalCount, right? But only for Astra, where theoretically a leader may be restarted and followers may keep going.  `perLevel` may have non-zero values that summed may be greater than the new totalCount, wdyt?

Reply:
It's still okay, isn't it? The code will simply do nothing until enough threads free up.

Reply:
Right, it's fine :+1: